Swap to just-stabilised asm!() and global_asm!() macros

adamgreig opened the pull request:
Once Rust 1.59 is released in a couple of days, the `asm!()` and `global_asm!()` macros will become stable, and we will no longer require the various precompiled binaries we've used until now in cortex-m and cortex-m-rt. cc #420.

This PR uses `asm!()` in cortex-m, and removes the inline-asm feature, since I anticipate this going into cortex-m 0.8 and so we don't need to leave it behind for compatibility. In various places the previous version would call an extern C method when built for the native target, which would only fail at link time; to preserve the ability to build on x86 I've either made the whole method require the `cortex_m` configuration, or where appropriate/convenient simply skipped the `asm!()` call.

This PR replaces the old gcc-preprocessed `asm.S` in cortex-m-rt with use of `global_asm!()`, although since you can't normally use `#[cfg(...)]` attributes with `global_asm!()`, there's also a slightly scary macro modified from one invented by @Dirbaio for a similar purpose. I considered putting the initialisation of LR behind an armv6m flag, but since we want to restore it after calling `__pre_init` it seemed better to just leave it the same on both targets. I added Cargo features to optionally set SP and VTOR at startup, which has been variously requested but would previously have required multiplicatively more pre-built binaries. Now: no problem. Relevant issues:

* https://github.com/rust-embedded/cortex-m-rt/issues/283
* https://github.com/rust-embedded/cortex-m-rt/issues/55
* https://github.com/rust-embedded/cortex-m-rt/issues/254
* https://github.com/rust-embedded/cortex-m-rt/issues/102
* https://github.com/rust-embedded/cortex-m-rt/pull/338

I've tested these on a couple of targets (and updated the CI): on the whole there's a small improvement in code size due to everyone getting inlined asm, especially in `cortex_m::interrupt::free()`.

The major downside is we bump our MSRV from 1.42 (March 2020) to 1.59 (Feb 2022). For cortex-m, I propose putting these changes in the upcoming 0.8 release (which is technically what the master branch is already on) and not backporting. For cortex-m-rt I'm not sure: we don't have any other pending breaking changes, so we could consider a patch release. Anyway, this PR doesn't commit to any particular releases, so we can decide that later.

For cortex-m-semihosting/panic-semihosting I think a patch release would be ideal, especially since we had to yank the last c-m-sh release due to conflicting prebuilt binaries (a problem that should now vanish).

Also tagging these issues that I think might also benefit from new inline asm:
* https://github.com/rust-embedded/cortex-m/issues/265
* https://github.com/rust-embedded/cortex-m/issues/215
* https://github.com/rust-embedded/cortex-m/issues/406

rust-highfive commented:
r? @thalesfragoso

(rust-highfive has picked a reviewer for you, use r? to override)

thejpster commented on cortex-m-rt/src/lib.rs:
```diff
@@ -431,8 +444,148 @@
 extern crate cortex_m_rt_macros as macros;
 
 use core::fmt;
+use core::arch::global_asm;
```
I hope this clippy lint goes away in 1.59.

thejpster commented on src/asm.rs:
```diff
@@ -160,15 +209,17 @@ pub fn ttat(addr: *mut u32) -> u32 {
 #[cfg(armv8m)]
 pub unsafe fn bx_ns(addr: u32) {
-    call_asm!(__bxns(addr: u32));
+    asm!("BXNS {}", in(reg) addr, options(nomem, nostack, preserves_flags));
```
Why is this one all-caps?

thejpster commented:
Looks OK to me. Very excited to see this land!

adamgreig commented:
I've fixed that uppercase BXNS, set rust-version in the Cargo.toml (except the unchanged panic-itm), bumped those Cargos to edition 2021 while we're at it, and removed outdated references to old Rust versions from cortex-m-rt docs (since we now require 1.59 anyway, and they were already outdated with 1.42).

thejpster commented:
What happened to the CI?

Dirbaio commented on Cargo.toml:
```diff
@@ -29,7 +30,6 @@ optional = true
 [features]
 cm7 = []
 cm7-r0p1 = ["cm7"]
-inline-asm = []
```
removing this is a breaking change, is the intent to release this as 0.8?

Perhaps it's better to leave it in with a note "deprecated, does nothing, kept for backwards compat"?

adamgreig replied:
Yes, from top of thread: "For cortex-m, I propose putting these changes in the upcoming 0.8 release (which is technically what the master branch is already on) and not backporting."

Dirbaio replied:
The `inline-asm` features in the other crates could be removed then?

adamgreig replied:
> For cortex-m-semihosting/panic-semihosting I think a patch release would be ideal, especially since we had to yank the last c-m-sh release due to conflicting prebuilt binaries (a problem that should now vanish).

But I'm sort of on the fence there, we could do remove the feature and do a breaking release. For cortex-m-semihosting we already released and yanked 0.4 and are bank on 0.3, so we could drop the feature and do 0.4 there. Perhaps that's nicest.

adamgreig replied:
Pushed an update removing it from those crates too, may as well make a clean sweep of it.

adamgreig commented:
Whew, a few snags from the CI but all green now.

therealprof reviewed (approved):
LGTM

adamgreig commented:
bors r=thejpster
